### PR TITLE
Add optional GITHUB_MERGE_BOT_TRUSTED_TEAMS_PERMISSION_SCOPE trusted teams approvals scope configuration

### PR DESCRIPTION
**Pull request by rymir**
When `GITHUB_MERGE_BOT_TRUSTED_TEAMS_PERMISSION_SCOPE` is set to the trusted repository owner teams permission scope (ex. admin, write),
after rebasing, `github-merge-bot` will reapprove only if the pull request
has been approved by a member of the teams with the configured permission scope.

### CODE REVIEW

File: test/github_merge_bot/core_test.clj [hunk 1]
```diff
@@ -69,10 +110,12 @@
         rev-c (git/git-add-and-commit origin-repo "C")
         _ (git/git-checkout origin-repo "master")
         repo (:repo (git/git-clone-full origin-repo-dir (str (System/getProperty "java.io.tmpdir") (UUID/randomUUID))))]
-    (with-redefs [procure-repo (constantly repo)]
-      (update-pull-request "foo" "bar" (pull-request-for-repo-branch origin-repo "feature-c") {:username "foo" :password "abc"})
+    (with-redefs [procure-repo (constantly repo)
+                  github-teams (constantly [{:name "team1" :id 1}])
+                  github-team-members (constantly [{:id 1}])]
+      (update-pull-request "foo" "team1" "bar" (pull-request-for-repo-branch origin-repo "feature-c") {:username "foo" :password "abc"})
```
Comment (pallamidessi) [inline test/github_merge_bot/core_test.clj]: I would suggest adding a test when no "team" envvar is set

File: src/github_merge_bot/core.clj [hunk 2]
```diff
@@ -17,14 +18,50 @@
...
 
-(defn approved? [owner repo pull-request]
-  (let [reviews (github-reviews owner repo (:number pull-request))
-        reviews-by-author (vals (group-by #(-> % :user :id) reviews))
-        decisive-reviews (map (fn [reviews]
-                                (->> reviews (map :state) (filter #{"CHANGES_REQUESTED" "APPROVED"}) last))
-                              reviews-by-author)]
-    (and (< 0 (count (filter #{"APPROVED"} decisive-reviews)))
-         (not-any? #{"CHANGES_REQUESTED"} decisive-reviews))))
+(defn github-teams [owner & [options]]
```
Comment (pallamidessi) [inline src/github_merge_bot/core.clj]: For all those new function and to be consistent with the rest of the codebase, they should have an action verb, i.e  `get-github-teams`, `group-reviews-by-authors`, `(filter | select)-reviews-by-team-member`

File: test/github_merge_bot/core_test.clj (hunk 1 above)
Comment (rymir) [reply test/github_merge_bot/core_test.clj]: https://github.com/depop/github-merge-bot/pull/1/files#diff-de2bedac227a2402a5110fdb6a9bb9d7R45

File: src/github_merge_bot/core.clj (hunk 2 above)
Comment (rymir) [reply src/github_merge_bot/core.clj]: done, kept the github ones `github-` as they were before

File: README.md [hunk 3]
```diff
@@ -40,12 +40,23 @@ Seeing as these projects did not fulfil our requirements, we decided to create g
...
+
+All github-merge-bot settings are configurable from the following environmental variables:
+
+Setting | Default | Description
+--- | --- | ---
+GITHUB_MERGE_BOT_OWNER | | The github organisation of the github repository github-merge-bot will integrate with
+GITHUB_MERGE_BOT_REPO | | The github repository github-merge-bot will integrate with
+GITHUB_MERGE_BOT_USERNAME | | The github user username github-merge-bot will integrate using
+GITHUB_MERGE_BOT_PASSWORD | | The github user password github-merge-bot will integrate using
+GITHUB_MERGE_BOT_REPO_OWNER_TEAM | nil | The github team which members approvals will be trusted when re-approving pull requests after a rebase operation. By default, any approvals will be trusted.
```
Comment (0x91) [inline README.md]: interesting this is a config, can we not infer it from the repo settings?

Comment (rymir) [reply README.md]: it's indeed a good point, I'll check if that's possible to get from the github apis

Comment (rymir) [reply README.md]: added a wip https://github.com/depop/github-merge-bot/pull/1/commits/5748e0b8f23ca48709fe657d1d3096ba52c95b24 where I refactor to use `GITHUB_MERGE_BOT_TRUSTED_TEAMS_PERMISSION_SCOPE`

Comment (rymir) [reply README.md]: based on `GITHUB_MERGE_BOT_TRUSTED_TEAMS_PERMISSION_SCOPE` I go and fetch the valid teams to consider for the pr re-approval

Comment (rymir) [reply README.md]: lemme know how this sounds, I can go and update the docs then

File: src/github_merge_bot/core.clj [hunk 4]
```diff
@@ -17,14 +18,39 @@
...
+(defn group-reviews-by-author [reviews]
+  (vals (group-by #(-> % :user :id) reviews)))
+
+(defn is-review-by-any-team-member? [review team-members]
+  (some (has-value :login (get (get review :user) :login)) team-members))
+
+(defn filter-reviews-by-team-member [reviews team-members trusted-team-permission-scope]
+  (match [reviews team-members trusted-team-permission-scope]
+    [reviews team-members nil] reviews
+    [reviews team-members trusted-team-permission-scope]
```
Comment (pallamidessi) [inline src/github_merge_bot/core.clj]: I don't think this is needed: we either have no `team-members` at all (so we just return `reviews`) or have some `team-member` and don't need the `trusted-team-permission-scope` var

i.e:
```
(defn filter-reviews-by-team-member [reviews team-members]
  (match [reviews team-members trusted-team-permission-scope]
    [reviews nil] reviews
    [reviews team-members]
      (filter #(is-review-by-any-team-member? % team-members) reviews)))
```
the member list would be nil anyway if the option is not set

Comment (pallamidessi) [reply src/github_merge_bot/core.clj]: if you want to keep it, I would suggest renaming it `trusted-team-permission-scope` -> `has-trusted-team-enabled` to make it more explicit ;)

Comment (rymir) [reply src/github_merge_bot/core.clj]: I would like to remove it, but we need it for the case where the `trusted-team-permission-scope` is unset, in that case, even though we've got no team-members, we want to return the full reviews list

Comment (0x91) [reply src/github_merge_bot/core.clj]: you could make it a multi-arity function then to make it clearer, not sure why we use match here:  https://clojure.org/guides/learn/functions#_multi_arity_functions

File: README.md (hunk 3 above)
Comment (0x91) [reply README.md]: I think this is good enough for now. this doesn't quite mirror the GitHub functionality to have required reviewers so you can configure it in one place and the bot will respect it, but I think it's an OK compromise.

**Comment by 0x91** [comment]
LGTM 👍

File: src/github_merge_bot/core.clj (hunk 4 above)
Comment (pallamidessi) [reply src/github_merge_bot/core.clj]: Use https://clojuredocs.org/clojure.string/blank_q :D

Comment (rymir) [reply src/github_merge_bot/core.clj]: @pallamidessi https://github.com/depop/github-merge-bot/pull/1/commits/2038b7fff66cca0ff695c03e469d34dbf66840e1

File: src/github_merge_bot/core.clj [hunk 5]
```diff
@@ -34,17 +34,16 @@
...
+  (if is-trusted-team-permission-scope-active
+    (filter #(is-review-by-any-team-member? % team-members) reviews)
+    reviews))
 
 (defn approved? [owner repo pull-request trusted-team-permission-scope]
    (let [repo-owners-teams (filter #(= (get % :permission) trusted-team-permission-scope) (github-repo-teams owner repo))
          repo-owners-teams-members (flatten (map #(github-team-members owner %) (map :id repo-owners-teams)))
          reviews (github-reviews owner repo (:number pull-request))
-         reviews-by-team-member (filter-reviews-by-team-member reviews repo-owners-teams-members trusted-team-permission-scope)
+         reviews-by-team-member (filter-reviews-by-team-member reviews repo-owners-teams-members (clojure.string/blank? trusted-team-permission-scope))
```
Comment (pallamidessi) [inline src/github_merge_bot/core.clj]: Sorry... Can you import it ?

Comment (rymir) [reply src/github_merge_bot/core.clj]: done!

**Pull request merged**